pkg/controllers/scheduling/schedule.go

```diff
@@ -17,42 +17,110 @@ import (
 	clusterlisterv1alpha1 "open-cluster-management.io/api/client/cluster/listers/cluster/v1alpha1"
 	clusterapiv1 "open-cluster-management.io/api/cluster/v1"
 	clusterapiv1alpha1 "open-cluster-management.io/api/cluster/v1alpha1"
+	"open-cluster-management.io/placement/pkg/plugins"
+	"open-cluster-management.io/placement/pkg/plugins/balance"
+	"open-cluster-management.io/placement/pkg/plugins/predicate"
+	"open-cluster-management.io/placement/pkg/plugins/steady"
 )
 
 const (
 	maxNumOfClusterDecisions = 100
 )
 
-type scheduleFunc func(
-	ctx context.Context,
-	placement *clusterapiv1alpha1.Placement,
-	clusters []*clusterapiv1.ManagedCluster,
-	clusterClient clusterclient.Interface,
-	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister,
-) (*scheduleResult, error)
+type Scheduler interface {
+	schedule(
+		ctx context.Context,
+		placement *clusterapiv1alpha1.Placement,
+		clusters []*clusterapiv1.ManagedCluster,
+		clusterClient clusterclient.Interface,
+		placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister,
+	) (*scheduleResult, error)
+}
 
 type scheduleResult struct {
 	feasibleClusters     int
 	scheduledDecisions   int
 	unscheduledDecisions int
 }
 
-func schedule(
+type pluginScheduler struct {
+	plugins []plugins.Plugin
+}
+
+func newPluginScheduler(handle plugins.Handle) *pluginScheduler {
+	return &pluginScheduler{
+		plugins: []plugins.Plugin{
+			predicate.New(handle),
+			steady.New(handle),
+			balance.New(handle),
+		},
+	}
+}
+
+func (s *pluginScheduler) schedule(
 	ctx context.Context,
 	placement *clusterapiv1alpha1.Placement,
 	clusters []*clusterapiv1.ManagedCluster,
 	clusterClient clusterclient.Interface,
 	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister,
 ) (*scheduleResult, error) {
-	// filter clusters with cluster predicates
-	feasibleClusters, err := matchWithClusterPredicates(placement.Spec.Predicates, clusters)
-	if err != nil {
-		return nil, err
+	var err error
+	filtered := clusters
+
+	// set plugins at first
+	for _, p := range s.plugins {
+		p.Set(placement)
 	}
 
+	// filter clusters
+	for _, p := range s.plugins {
+		filter, ok := p.(plugins.Filter)
+		if !ok {
+			continue
+		}
+		filtered, err = filter.Filter(ctx, filtered)
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// score clusters
+	// Score the cluster
+	scoreSum := map[string]float64{}
+	for _, cluster := range filtered {
+		scoreSum[cluster.Name] = 0.0
+	}
+	for _, p := range s.plugins {
+		prioritizer, ok := p.(plugins.Prioritizer)
+		if !ok {
+			continue
+		}
+		score, err := prioritizer.Score(ctx, filtered)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO we currently weigh each prioritizer as equal. We should consider
+		// importance factor for each priotizer when caculating the final score.
```

Comment:
This is one way to do it.  Another way to consider it is to say that individual prioritizers have configuration that cause them to weight differently.